File: sdk/go/request_boundary_test.go
package nextloggers

import (
	"context"
	"errors"
	"fmt"
	"sync"
	"testing"
	"time"
)

func boundaryTestContext(slot int) RequestContext {
	return RequestContext{
		RequestID:      fmt.Sprintf("request-%d", slot),
		LoggedInUserID: fmt.Sprintf("user-%d", slot),
		TenantID:       fmt.Sprintf("tenant-%d", slot),
		TraceID:        fmt.Sprintf("%032x", slot),
	}
}

func boundaryForSlot(slot int) RequestBoundary {
	switch slot % 5 {
	case 0:
		return HTTPRequestBoundary("handler", fmt.Sprintf("http-%d", slot))
	case 1:
		return TCPConnectionBoundary("accept", fmt.Sprintf("connection-%d", slot), "tcp.accept")
	case 2:
		return TCPMessageBoundary("decode", fmt.Sprintf("connection-%d", slot), fmt.Sprintf("message-%d", slot), "tcp.decode")
	case 3:
		return WebSocketSessionBoundary("upgrade", fmt.Sprintf("session-%d", slot), "websocket.upgrade")
	case 4:
		return WebSocketMessageBoundary("dispatch", fmt.Sprintf("session-%d", slot), fmt.Sprintf("message-%d", slot), "websocket.dispatch")
	default:
		panic("unreachable slot")
	}
}

func TestRequestBoundariesKeepConcurrentFailuresIsolated(t *testing.T) {
	t.Parallel()
	const count = 64
	type report struct {
		requestID string
		transport RequestTransport
		scope     RequestScope
	}
	reports := make([]report, 0, count)
	var reportsMu sync.Mutex
	results := make([]RequestBoundaryResult[int], count)

	var group sync.WaitGroup
	group.Add(count)
	for slot := 0; slot < count; slot++ {
		slot := slot
		go func() {
			defer group.Done()
			results[slot] = RunWithRequestBoundary(
				context.Background(),
				boundaryTestContext(slot),
				boundaryForSlot(slot),
				func(ctx context.Context) (int, error) {
					time.Sleep(time.Duration((count-slot)%7) * time.Millisecond)
					requestID, ok := RequestIDFrom(ctx)
					if !ok || requestID != fmt.Sprintf("request-%d", slot) {
						return 0, fmt.Errorf("wrong request context: %q", requestID)
					}
					return 0, fmt.Errorf("failure-%d", slot)
				},
				RequestBoundaryOptions{
					Now: func() time.Time { return time.UnixMilli(int64(1_000 + slot)) },
					Report: func(ctx context.Context, failure RequestBoundaryFailure) {
						requestID, ok := RequestIDFrom(ctx)
						if !ok || requestID != fmt.Sprintf("request-%d", slot) {
							panic(fmt.Sprintf("reporter saw wrong request: %q", requestID))
						}
						reportsMu.Lock()
						reports = append(reports, report{
							requestID: failure.Context.RequestID,
							transport: failure.Boundary.Transport,
							scope:     failure.Boundary.Scope,
						})
						reportsMu.Unlock()
					},
				},
			)
		}()
	}
	group.Wait()

	if len(reports) != count {
		t.Fatalf("expected %d reports, got %d", count, len(reports))
	}
	for slot, result := range results {
		if result.OK() || result.Failure == nil {
			t.Fatalf("slot %d unexpectedly succeeded", slot)
		}
		failure := result.Failure
		if failure.Kind != RequestFailureException {
			t.Fatalf("slot %d: unexpected kind %q", slot, failure.Kind)
		}
		if got, want := failure.Context.RequestID, fmt.Sprintf("request-%d", slot); got != want {
			t.Fatalf("slot %d: request ID = %q, want %q", slot, got, want)
		}
		if got, want := failure.ObservedAtUnixMS, int64(1_000+slot); got != want {
			t.Fatalf("slot %d: observed time = %d, want %d", slot, got, want)
		}
		if failure.Err == nil || failure.Err.Error() != fmt.Sprintf("failure-%d", slot) {
			t.Fatalf("slot %d: wrong error: %v", slot, failure.Err)
		}
	}
}

func TestRequestBoundaryRecoversOperationAndReporterPanics(t *testing.T) {
	t.Parallel()
	result := RunWithRequestBoundary(
		context.Background(),
		boundaryTestContext(70),
		WebSocketMessageBoundary("dispatch", "session-70", "message-70", "websocket.dispatch"),
		func(context.Context) (int, error) {
			panic("handler panic")
		},
		RequestBoundaryOptions{
			Report: func(context.Context, RequestBoundaryFailure) {
				panic("telemetry panic")
			},
		},
	)

	if result.Failure == nil || result.Failure.Kind != RequestFailurePanic {
		t.Fatalf("expected recovered panic, got %#v", result.Failure)
	}
	if result.Failure.Recovered != "handler panic" {
		t.Fatalf("wrong recovered value: %#v", result.Failure.Recovered)
	}
	if result.Failure.Context.RequestID != "request-70" {
		t.Fatalf("panic lost request context: %#v", result.Failure.Context)
	}
}

func TestRequestBoundaryClassifiesCancellationTimeoutAndDisconnect(t *testing.T) {
	t.Parallel()
	tests := []struct {
		name string
		err  error
		want RequestFailureKind
	}{
		{name: "timeout", err: context.DeadlineExceeded, want: RequestFailureTimeout},
		{name: "cancelled", err: context.Canceled, want: RequestFailureCancelled},
		{name: "disconnect", err: fmt.Errorf("socket closed: %w", ErrPeerDisconnected), want: RequestFailureDisconnect},
	}

	for index, testCase := range tests {
		testCase := testCase
		t.Run(testCase.name, func(t *testing.T) {
			t.Parallel()
			result := RunWithRequestBoundary(
				context.Background(),
				boundaryTestContext(80+index),
				TCPMessageBoundary("read", "connection", "message", "tcp.read"),
				func(context.Context) (int, error) { return 0, testCase.err },
				RequestBoundaryOptions{},
			)
			if result.Failure == nil || result.Failure.Kind != testCase.want {
				t.Fatalf("kind = %#v, want %q", result.Failure, testCase.want)
			}
		})
	}
}

func TestRequestBoundaryRestoresExplicitParentContext(t *testing.T) {
	t.Parallel()
	parent := WithRequestContext(context.Background(), boundaryTestContext(90))
	result := RunWithRequestBoundary(
		parent,
		boundaryTestContext(91),
		HTTPRequestBoundary("handler", "http.handler"),
		func(ctx context.Context) (string, error) {
			requestID, _ := RequestIDFrom(ctx)
			return requestID, nil
		},
		RequestBoundaryOptions{},
	)
	if !result.OK() || result.Value != "request-91" {
		t.Fatalf("unexpected result: %#v", result)
	}
	requestID, ok := RequestIDFrom(parent)
	if !ok || requestID != "request-90" {
		t.Fatalf("parent context changed: %q", requestID)
	}
}

func TestInvalidBoundaryFailsWithoutRunningOperation(t *testing.T) {
	t.Parallel()
	ran := false
	result := RunWithRequestBoundary(
		context.Background(),
		boundaryTestContext(100),
		RequestBoundary{Transport: RequestTransportHTTP, Scope: RequestScopeMessage, Phase: "handler"},
		func(context.Context) (int, error) {
			ran = true
			return 1, nil
		},
		RequestBoundaryOptions{},
	)
	if ran {
		t.Fatal("invalid boundary executed the operation")
	}
	if result.Failure == nil || !errors.Is(result.Failure.Err, result.Failure.Err) {
		t.Fatalf("expected validation failure, got %#v", result.Failure)
	}
	if result.Failure.Context.RequestID != "request-100" {
		t.Fatalf("validation failure lost request context: %#v", result.Failure.Context)
	}
}

File: tests/request-boundary.test.mjs
import assert from 'node:assert/strict';
import test from 'node:test';

import {
  getLogContext,
  httpRequestBoundary,
  runWithLogContext,
  runWithRequestBoundary,
  tcpConnectionBoundary,
  tcpMessageBoundary,
  webSocketMessageBoundary,
  webSocketSessionBoundary,
} from '../dist/context.js';
import {
  getLogContext as getBrowserLogContext,
  runWithRequestBoundary as runWithBrowserBoundary,
} from '../dist/context-browser.js';
import {
  getLogContext as getWorkerdLogContext,
  runWithRequestBoundary as runWithWorkerdBoundary,
} from '../dist/context-workerd.js';

const delay = (milliseconds) =>
  new Promise((resolve) => setTimeout(resolve, milliseconds));

function contextFor(slot) {
  return {
    traceId: slot.toString(16).padStart(32, '0'),
    traceIds: [slot.toString(16).padStart(32, '0')],
    routineId: `request-${slot}`,
    fields: {
      'request.id': `request-${slot}`,
      'user.id': `user-${slot}`,
      'tenant.id': `tenant-${slot}`,
    },
    tags: ['request-boundary-test'],
  };
}

function boundaryFor(slot) {
  switch (slot % 5) {
    case 0:
      return httpRequestBoundary('handler', `http-${slot}`);
    case 1:
      return tcpConnectionBoundary('accept', `connection-${slot}`);
    case 2:
      return tcpMessageBoundary(
        'decode',
        `connection-${slot}`,
        `message-${slot}`,
      );
    case 3:
      return webSocketSessionBoundary('upgrade', `session-${slot}`);
    case 4:
      return webSocketMessageBoundary(
        'dispatch',
        `session-${slot}`,
        `message-${slot}`,
      );
    default:
      throw new Error('unreachable slot');
  }
}

test('parallel HTTP, TCP, and WebSocket failures remain pinned to one request', async () => {
  const reports = [];
  const count = 60;
  const results = await Promise.all(
    Array.from({ length: count }, async (_, slot) =>
      runWithRequestBoundary(
        contextFor(slot),
        boundaryFor(slot),
        async () => {
          await delay((count - slot) % 9);
          assert.equal(
            getLogContext()?.fields?.['request.id'],
            `request-${slot}`,
          );
          throw new Error(`failure-${slot}`);
        },
        {
          now: () => 1_000 + slot,
          report: async (failure) => {
            assert.equal(
              getLogContext()?.fields?.['request.id'],
              `request-${slot}`,
            );
            await delay(slot % 4);
            reports.push({
              requestId: failure.context.fields?.['request.id'],
              transport: failure.boundary.transport,
              scope: failure.boundary.scope,
              observedAtUnixMs: failure.observedAtUnixMs,
            });
          },
        },
      ),
    ),
  );

  assert.equal(getLogContext(), undefined);
  assert.equal(reports.length, count);
  assert.equal(results.length, count);
  for (let slot = 0; slot < count; slot += 1) {
    const result = results[slot];
    assert.equal(result.ok, false);
    assert.equal(result.failure.kind, 'exception');
    assert.equal(
      result.failure.context.fields?.['request.id'],
      `request-${slot}`,
    );
    assert.equal(result.failure.observedAtUnixMs, 1_000 + slot);
    assert.equal(result.failure.cause.message, `failure-${slot}`);
  }
});

test('nested failure boundaries restore the parent frame', async () => {
  const outer = contextFor(100);
  const inner = contextFor(101);

  await runWithLogContext(outer, async () => {
    assert.equal(getLogContext()?.fields?.['request.id'], 'request-100');
    const result = await runWithRequestBoundary(
      inner,
      webSocketMessageBoundary('dispatch', 'session-101', 'message-101'),
      async () => {
        assert.equal(getLogContext()?.fields?.['request.id'], 'request-101');
        throw new Error('inner failure');
      },
      {
        report(failure) {
          assert.equal(
            getLogContext()?.fields?.['request.id'],
            'request-101',
          );
          assert.equal(
            failure.context.fields?.['request.id'],
            'request-101',
          );
        },
      },
    );

    assert.equal(result.ok, false);
    assert.equal(getLogContext()?.fields?.['request.id'], 'request-100');
  });

  assert.equal(getLogContext(), undefined);
});

test('reporter failure cannot replace the original operation failure', async () => {
  const original = new Error('original request failure');
  const result = await runWithRequestBoundary(
    contextFor(200),
    httpRequestBoundary('finalize'),
    async () => {
      throw original;
    },
    {
      classify: () => 'timeout',
      report: async () => {
        throw new Error('telemetry sink unavailable');
      },
    },
  );

  assert.equal(result.ok, false);
  assert.equal(result.failure.kind, 'timeout');
  assert.equal(result.failure.cause, original);
  assert.equal(getLogContext(), undefined);
});

test('successful boundaries preserve values and do not report failures', async () => {
  let reports = 0;
  const result = await runWithRequestBoundary(
    contextFor(300),
    tcpMessageBoundary('dispatch', 'connection-300', 'message-300'),
    async () => {
      await delay(1);
      return 42;
    },
    { report: () => { reports += 1; } },
  );

  assert.deepEqual(result, { ok: true, value: 42 });
  assert.equal(reports, 0);
  assert.equal(getLogContext(), undefined);
});

test('browser fallback carries explicit failure context without async ambient bleed', async () => {
  const count = 24;
  const results = await Promise.all(
    Array.from({ length: count }, async (_, slot) =>
      runWithBrowserBoundary(
        contextFor(slot),
        webSocketMessageBoundary(
          'browser-dispatch',
          `session-${slot}`,
          `message-${slot}`,
        ),
        async () => {
          await delay((count - slot) % 5);
          assert.equal(getBrowserLogContext(), undefined);
          throw new Error(`browser-${slot}`);
        },
      ),
    ),
  );

  assert.equal(getBrowserLogContext(), undefined);
  for (let slot = 0; slot < count; slot += 1) {
    assert.equal(results[slot].ok, false);
    assert.equal(
      results[slot].failure.context.fields?.['request.id'],
      `request-${slot}`,
    );
  }
});

test('workerd explicit-only fallback never exposes ambient request identity', async () => {
  const result = await runWithWorkerdBoundary(
    contextFor(400),
    tcpConnectionBoundary('workerd-connection', 'connection-400'),
    async () => {
      assert.equal(getWorkerdLogContext(), undefined);
      await delay(1);
      assert.equal(getWorkerdLogContext(), undefined);
      throw new Error('workerd failure');
    },
  );

  assert.equal(result.ok, false);
  assert.equal(
    result.failure.context.fields?.['request.id'],
    'request-400',
  );
  assert.equal(getWorkerdLogContext(), undefined);
});
